lib/logstash/pipeline.rb

```diff
@@ -183,14 +188,23 @@ def inputworker(plugin)
                              :plugin => plugin.inspect, :error => e))
       end
       puts e.backtrace if @logger.debug?
-      plugin.teardown
+      # input teardown must be synchronized since is can be called concurrently by
+      # the input worker thread and from the pipeline thread shutdown method.
+      # this means that input teardown methods must support multiple calls.
+      @run_mutex.synchronize{plugin.teardown}
```

Comment:
upon shutdown, most of the times, teardown is actually called by two thread concurrently, at this point, without doing too much refactoring, the least we can do is to prevent this 2 teardown calls to run *at the same time*.